dashboard/app/models/school_stats_by_year.rb

```diff
@@ -82,7 +55,11 @@ def self.merge_from_csv(filename, options = {col_sep: "\t", headers: true, quote
       parsed = yield row
       loaded = find_by(primary_keys.map(&:to_sym).map {|k| [k, parsed[k]]}.to_h)
       if loaded.nil?
-        SchoolStatsByYear.new(parsed).save!
+        begin
+          SchoolStatsByYear.new(parsed).save!
+        rescue ActiveRecord::InvalidForeignKey
+          puts parsed[:school_id]
```

Comment:
Before, an exception raised here would roll back the transaction.  Now that you've added a `rescue` block that's not the case.  Is that by design?

If we're going to leave this `puts` statement in, it might be good to make it more descriptive.

Reply:
I'm kind of surprised the previous script successfully ran without anything to catch exceptions, our `schools` table is such a mishmash of various imports from different years that typically there are some records that don't match between various datasets, etc.

But yes, it's intentional. Caught 230 schools (out of 100K) with no entry in `schools`. I guess it'd be nice to make sure that the number of schools being caught isn't something extraordinary (I'd be ok with a 1% error rate, so maybe limit to 1000)?

Forgot to change the `puts` to something more descriptive -- there's a better way to do this anyway, right? Like CDO.log or something?

Reply:
:+1: Sounds good. Re: logging, for a oneoff script `puts` should be fine, since it'll always be attended while running and you can capture the output however you'd like.

Reply:
Did a little exploration of why this might be happening, notes for posterity:

I looked at a school that was skipped (named AULT EL, which is in Texas and has a NCES ID of 481611016970). Looking at the raw CSV in S3, this school was in the list of schools to import when we imported schools from the [2017-18 iteration of the NCES database](https://github.com/code-dot-org/code-dot-org/pull/25707) last year, but was skipped because it already exists in our schools list with a very similar NCES ID (481611006970) and the same `state_school_id` (a state's identifier for a school -- TX-101907-101907133). It looks like we [enforce uniqueness](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/school.rb#L26) of the `state_school_id` in addition to the NCES `school_id` (ID column in the `schools` table), so I could see why it got skipped.

Reply:
Another note for the record -- we added the uniqueness requirement for `state_school_id` after the most recent NCES school demographic info was imported, so I think that would explain why there wouldn't have been any foreign key errors in the past.